crates/payload/basic/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+        }
+    }
+}
+
+// === impl BasicPayloadJobGenerator ===
+
+impl<Client, Pool, Tasks> BasicPayloadJobGenerator<Client, Pool, Tasks> {}
+
+impl<Client, Pool, Tasks> PayloadJobGenerator for BasicPayloadJobGenerator<Client, Pool, Tasks>
+where
+    Client: StateProviderFactory + BlockProvider + Clone + Unpin + 'static,
+    Pool: TransactionPool + Unpin + 'static,
+    Tasks: TaskSpawner + Clone + Unpin + 'static,
+{
+    type Job = BasicPayloadJob<Client, Pool, Tasks>;
+
+    fn new_payload_job(
+        &self,
+        attributes: PayloadBuilderAttributes,
+    ) -> Result<Self::Job, PayloadBuilderError> {
+        // TODO this needs to access the _pending_ state of the parent block hash
+        let parent_block = self
+            .client
+            .block_by_hash(attributes.parent)?
+            .ok_or_else(|| PayloadBuilderError::MissingParentBlock(attributes.parent))?;
+
+        // configure evm env based on parent block
+        let initialized_cfg = CfgEnv {
+            chain_id: U256::from(self.chain_spec.chain().id()),
+            // ensure we're not missing any timestamp based hardforks
+            spec_id: revm_spec_by_timestamp_after_merge(&self.chain_spec, attributes.timestamp),
+            ..Default::default()
+        };
+
+        let initialized_block_env = BlockEnv {
+            number: U256::from(parent_block.number + 1),
+            coinbase: attributes.suggested_fee_recipient,
+            timestamp: U256::from(attributes.timestamp),
+            difficulty: U256::ZERO,
+            prevrandao: Some(attributes.prev_randao),
+            gas_limit: U256::from(parent_block.gas_limit),
+            // calculate basefee based on parent block's gas usage
+            basefee: U256::from(calculate_next_block_base_fee(
+                parent_block.gas_used,
+                parent_block.gas_limit,
+                parent_block.base_fee_per_gas.unwrap_or_default(),
+            )),
+        };
+
+        let config = PayloadConfig {
+            initialized_block_env,
+            initialized_cfg,
+            parent_block: Arc::new(parent_block),
+            extra_data: self.block_config.extradata.clone(),
+            attributes,
+            chain_spec: Arc::clone(&self.chain_spec),
+        };
+
+        // create empty
+
+        let until = tokio::time::Instant::now() + self.config.deadline;
+        let deadline = Box::pin(tokio::time::sleep_until(until));
+
+        Ok(BasicPayloadJob {
+            config,
+            client: self.client.clone(),
+            pool: self.pool.clone(),
+            executor: self.executor.clone(),
+            deadline,
+            interval: tokio::time::interval(self.config.interval),
+            best_payload: None,
+            pending_block: None,
+            payload_task_guard: self.payload_task_guard.clone(),
+        })
+    }
+}
+
+/// Restricts how many generator tasks can be executed at once.
+#[derive(Clone)]
+struct PayloadTaskGuard(Arc<Semaphore>);
```

Comment:
really nice pattern, in general will be really interesting to hear you explain / talk payload building. also maybe we should combine payload and miner into some top-level crate? or put them in a shared subdir?